### ISSUE TITLE
Check for issues with dice and focal losses

### ISSUE THREAD

**Issue by samvanstroud** [samvanstroud/hepattn#44]
Dice loss looks not good

<img width="1502" alt="Image" src="https://github.com/user-attachments/assets/b54c222e-86d2-4b5c-b28b-0a309bd02c89" />

**Comment by samvanstroud** [comment]
another longer example 

<img width="1505" alt="Image" src="https://github.com/user-attachments/assets/3768b82e-7969-44f0-819b-bd9860611f5f" />

**Comment by dkobylianskii** [comment]
At my branch, it looks fine

<img width="1492" alt="Image" src="https://github.com/user-attachments/assets/0e59ef6f-3f6b-4119-8005-edfddbfc4432" />

**Comment by samvanstroud** [comment]
Good to know, thanks @dkobylianskii! I willl try and pull your changes to the losses into main. cc @trahxam @sunnyw22

**Comment by sunnyw22** [comment]
In the main repo, the `num_pred` for the dice and iou cost appears to be missing a sigmoid function! (i.e. num_pred = pred_logits.sigmoid().sum(-1).unsqueeze(-1))

**Comment by samvanstroud** [comment]
Having discussed a bit more with @sunnyw22 and @trahxam, I think there are a few issues related to masking:

1. We don't pass any masking information to the costs [here](https://github.com/samvanstroud/hepattn/blob/868a40e29bb24954e83a826c75671f38f26140d6/src/hepattn/models/task.py#L265)
2. Different losses (dice vs CE/focal) need different approaches to masking queries versus inputs. We should pass the object and hit valid masks separately for each loss and let them handle the masking of each
